### DISCUSSION TITLE
Need help with fetching events from the EG Fortnite API

### DISCUSSION THREAD

**Discussion by iitsg** [LeleDerGrasshalmi/FortniteEndpointsDocumentation#320] [Q&A] [answered]
I am currently trying to understand the Epic Games API, and the documentation does help, but not completely. I am a bit confused about what to enter and whether I did everything correctly up until now.

My current code, written in Python (without imports or redundant Python-necessary stuff to reduce clutter):
```py
# Fortnite PC Game Client credentials from egs.jaren.wtf
client_id = "REDACTED_CLIENT_ID"
client_secret = "REDACTED_CLIENT_SECRET"

credentials = f"{client_id}:{client_secret}"
encoded_credentials = base64.b64encode(credentials.encode()).decode()


def get_access_token():
    token_url = (
        "https://account-public-service-prod03.ol.epicgames.com/account/api/oauth/token"
    )
    headers = {
        "Content-Type": "application/x-www-form-urlencoded",
        "Authorization": f"Basic {encoded_credentials}",
    }
    data = {
        "grant_type": "client_credentials",
    }
    response = requests.post(token_url, headers=headers, data=data)
    response_data = response.json()

    if "access_token" in response_data:
        return response_data["access_token"]
    else:
        raise Exception(f"Failed to obtain access token: {response_data}")


def fetch_events(access_token):
    game_id = "Fortnite"
    account_id = "REDACTED_ACCOUNT_ID" # I just entered my Epic Account ID from Fortnite settings ¯\_(ツ)_/¯
    event_url = f"https://events-public-service-live.ol.epicgames.com/api/v1/events/{game_id}/download/{account_id}"
    headers = {
        "Authorization": f"Bearer {access_token}",
        "Content-Type": "application/json",
    }

    region = "REDACTED_REGION" 
    platform = "REDACTED_PLATFORM"

    params = {"region": region, "platform": platform, "teamAccountIds": [account_id]} # Is this correct as well? What are these "teamAccountIds" anyway?

    response = requests.get(event_url, headers=headers, params=params)

    if response.status_code == 200:
        return response.json()
    else:
        raise Exception(
            f"Failed to fetch events: {response.status_code} - {response.text}"
        )
```

What I want to achieve is to gather all upcoming events planned (shown in the Compete tab in Fortnite), without my own account's data attached. But I only get this error back:
```vbnet
An error occurred: Failed to fetch events: 403 - {"errorCode":"errors.com.epicgames.common.forbidden","errorMessage":"Sorry your login does not posses the permissions 'Fortnite:profile:REDACTED_ACCOUNT_ID:commands READ' needed to perform the requested operation"}
```
I thought the Fortnite PC Game Client credentials had the correct permissions though:
![image](https://github.com/user-attachments/assets/4acb909c-a1af-41b5-bf1a-6a231a2d56ea)

Can anyone help me get this to work? I really want to get this working, as the API seems to be the only reliable and detailed source for the information I want/need.

**Comment by LeleDerGrasshalmi** [comment]
hi,

1. why redact public client id and secret?
2. infact, you need account authentication to use that api, you are using client auth

**Reply by iitsg** [reply]
1. I just redacted everything, just to be sure I don't do any mistake. Rather safe, than sorry.
2. What do you mean by that?

**Reply by LeleDerGrasshalmi** [reply]
to 2. you are using client authentication, not tied to any account.

but you are trying to perform a user action, so you need to be authenticated as that account.

check out the [getting started guide](https://github.com/LeleDerGrasshalmi/FortniteEndpointsDocumentation/blob/main/EpicGames/AccountService/Authentication/README.md#getting-started) in this repo.

**Reply by iitsg** [reply]
Hey there!
I finally understood how this works and I fixed it. I was able to get the data I wanted, but for each request I needed to request a new `authorizationCode` as well. Is there a way to automate this, so it's also able to run automatically on a pi?
Since I need to be logged in at the same time, I may have to build something manually, if no easy solution is available.

Or is there an other way to access the event data, utilizing client auth?

**Reply by LeleDerGrasshalmi** [reply]
hi, you dont need to get a new authorization code each time. the token lasts for 2 hours, once that expires use the refresh tokens (lasts for 8h)

**Reply by iitsg** [reply]
Do I then send the refresh token back to the same endpoint for the access token, like documented here: [GrantType refresh_token](https://github.com/LeleDerGrasshalmi/FortniteEndpointsDocumentation/blob/main/EpicGames/AccountService/Authentication/GrantTypes/refresh_token.md)
And, if I keep the refresh token alive to get back another access token, can I do this infinitely, without any problems?

**Reply by LeleDerGrasshalmi** [reply, marked as answer]
yes, thats how oauth2 session lifetime is supposed to work.

however if both tokens expire you need to begin again from the start (authorization code input)
